maven-surefire-common/src/test/java/org/apache/maven/plugin/surefire/AbstractSurefireMojoToolchainsTest.java

```diff
@@ -37,33 +37,91 @@
 
 import static java.util.Collections.emptyMap;
 import static java.util.Collections.singletonList;
+import static java.util.Collections.singletonMap;
+import static junit.framework.TestCase.assertNull;
 import static org.fest.assertions.Assertions.assertThat;
-import static org.fest.assertions.Fail.fail;
 import static org.powermock.api.mockito.PowerMockito.mock;
+import static org.powermock.api.mockito.PowerMockito.mockStatic;
+import static org.powermock.api.mockito.PowerMockito.when;
 import static org.powermock.reflect.Whitebox.invokeMethod;
 
 /**
- * Test for {@link AbstractSurefireMojo}.
+ * Test for {@link AbstractSurefireMojo}. jdkToolchain parameter
  */
 @RunWith( PowerMockRunner.class )
 @PrepareForTest( {AbstractSurefireMojo.class, ResolvePathsRequest.class, ReflectionUtils.class} )
 @PowerMockIgnore( {"org.jacoco.agent.rt.*", "com.vladium.emma.rt.*"} )
 public class AbstractSurefireMojoToolchainsTest
 {
 
+    @Test
+    public void shouldCallMaven33xMethodWhenSpecSet() throws Exception
+    {
+        AbstractSurefireMojoTest.Mojo mojo = new AbstractSurefireMojoTest.Mojo();
+        Toolchain expectedFromMaven33Method = mock( Toolchain.class );
+        MockToolchainManager toolchainManager = new MockToolchainManager( null, null );
+        mojo.setToolchainManager( toolchainManager );
+        mojo.setJdkToolchain( singletonMap( "version", "1.8" ) );
+
+        mockStatic( AbstractSurefireMojo.class );
+        when(
+            AbstractSurefireMojo.class,
+            "getToolchainMaven33x",
+            ToolchainManager.class,
+            toolchainManager,
+            mojo.getSession(), mojo.getJdkToolchain() ).thenReturn( expectedFromMaven33Method );
+        Toolchain actual = invokeMethod( mojo, "getToolchain" );
+        assertThat( actual )
+            .isSameAs( expectedFromMaven33Method );
+    }
+
+    @Test
+    public void shouldFallthroughToBuildContextWhenNoSpecSet() throws Exception
+    {
+        AbstractSurefireMojoTest.Mojo mojo = new AbstractSurefireMojoTest.Mojo();
+        Toolchain expectedFromContext = mock( Toolchain.class );
+        Toolchain expectedFromSpec = mock( Toolchain.class ); //ensure it still behaves correctly even if not null
+        mojo.setToolchainManager( new MockToolchainManager( expectedFromSpec, expectedFromContext ) );
+        Toolchain actual = invokeMethod( mojo, "getToolchain" );
```

Comment:
According to the implementation of `MockToolchainManager`, i would say that `getToolchains` returns `expectedFromSpec`. Pls explain to me why it is `expectedFromContext`.

Reply:
Because jdkToolchain is null.  I added some javadoc to the test for anyone trying to figure this out in the future.